node/cmd/guardiand/node.go

```diff
@@ -771,36 +772,89 @@ func runNode(cmd *cobra.Command, args []string) {
 
-	// Outbound gossip message queue
-	sendC := make(chan []byte)
+		signedInReadC  <-chan *gossipv1.SignedVAAWithQuorum
+		signedInWriteC chan<- *gossipv1.SignedVAAWithQuorum
 
-	// Inbound observations
-	obsvC := make(chan *gossipv1.SignedObservation, 50)
+		obsvReqReadC  <-chan *gossipv1.ObservationRequest
+		obsvReqWriteC chan<- *gossipv1.ObservationRequest
 
-	// Inbound signed VAAs
-	signedInC := make(chan *gossipv1.SignedVAAWithQuorum, 50)
+		obsvReqSendReadC  <-chan *gossipv1.ObservationRequest
+		obsvReqSendWriteC chan<- *gossipv1.ObservationRequest
 
-	// Inbound observation requests from the p2p service (for all chains)
-	obsvReqC := make(chan *gossipv1.ObservationRequest, common.ObsvReqChannelSize)
+		injectReadC  <-chan *vaa.VAA
+		injectWriteC chan<- *vaa.VAA
+	)
 
-	// Outbound observation requests
-	obsvReqSendC := make(chan *gossipv1.ObservationRequest, common.ObsvReqChannelSize)
+	// Outbound gossip message queue (needs to be read/write because p2p needs read/write)
+	gossipSendC := make(chan []byte)
+	// Inbound observations
+	obsvC := make(chan *gossipv1.SignedObservation, 50)
 
-	// Injected VAAs (manually generated rather than created via observation)
-	injectC := make(chan *vaa.VAA)
+	// the following channels are created in a block such that they can only be accessed through their explicit read/write aliases
+	{
+		// Finalized guardian observations aggregated across all chains
+		msgC := make(chan *common.MessagePublication)
+		msgReadC = msgC
+		msgWriteC = msgC
+
+		// Ethereum incoming guardian set updates
+		setC := make(chan *common.GuardianSet)
+		setWriteC = setC
+
+		// Inbound observation requests from the p2p service (for all chains)
+		obsvReqC := make(chan *gossipv1.ObservationRequest, common.ObsvReqChannelSize)
+		obsvReqReadC = obsvReqC
+		obsvReqWriteC = obsvReqC
+
+		// Outbound observation requests
+		obsvReqSendC := make(chan *gossipv1.ObservationRequest, common.ObsvReqChannelSize)
+		obsvReqSendReadC = obsvReqSendC
+		obsvReqSendWriteC = obsvReqSendC
+
+		// Injected VAAs (manually generated rather than created via observation)
+		injectC := make(chan *vaa.VAA)
+		injectReadC = injectC
+		injectWriteC = injectC
+	}
 
 	// Guardian set state managed by processor
 	gst := common.NewGuardianSetState(nil)
 
 	// Per-chain observation requests
 	chainObsvReqC := make(map[vaa.ChainID]chan *gossipv1.ObservationRequest)
 
+	// Per-chain msgC
+	chainMsgC := make(map[vaa.ChainID]chan *common.MessagePublication)
+	// aggregate per-chain msgC into msgC.
+	// SECURITY defense-in-depth: This way we enforce that a watcher must set the msg.EmiggerChain to its chainId.
+	for chainId := range sdk.KnownTokenbridgeEmitters { // TODO maybe not use KnownTokenbridgeEmitters here and think about adding just a list of chainIds to the sdk mainnet_consts.go
+		chainMsgC[chainId] = make(chan *common.MessagePublication)
+		go func(c <-chan *common.MessagePublication, chainId vaa.ChainID) {
+			for msg := range c {
```

Comment:
This will leak the goroutine, which isn't that big of an issue but if we want to do it the Right Way (TM) then we should select on the root context here:

```go
for {
	select {
		case <-rootCtx.Done():
			return
		case msg := <-c:
			// Handle msg
	}
}
```